File: java/exercises/practice/unit_test_7b23bd3e_1047_4456_be54_3246884fac06/task_scheduler/src/test/java/TaskSchedulerTest.java
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import org.junit.jupiter.api.Test;

public class TaskSchedulerTest {

    // Helper task class used for testing
    // It is assumed that the main solution uses a similar Task definition.
    static class Task {
        int id;
        int duration;
        int deadline;
        List<Integer> dependencies;

        Task(int id, int duration, int deadline, List<Integer> dependencies) {
            this.id = id;
            this.duration = duration;
            this.deadline = deadline;
            this.dependencies = dependencies;
        }
    }

    // It is assumed that the main solution has a method with the following signature:
    // public static int minPenalty(List<Task> tasks)
    // which returns the minimum total penalty or Integer.MAX_VALUE if no valid schedule exists.

    // Test 1: Empty Task List
    @Test
    public void testEmptyTaskList() {
        List<Task> tasks = new ArrayList<>();
        int penalty = TaskScheduler.minPenalty(tasks);
        assertEquals(0, penalty, "Empty task list should incur 0 penalty.");
    }

    // Test 2: Single Task that finishes before deadline
    @Test
    public void testSingleTaskOnTime() {
        Task task = new Task(1, 10, 20, new ArrayList<>());
        List<Task> tasks = Collections.singletonList(task);
        int penalty = TaskScheduler.minPenalty(tasks);
        assertEquals(0, penalty, "Single task finishing on time should incur 0 penalty.");
    }

    // Test 3: Single Task that finishes after deadline
    @Test
    public void testSingleTaskLate() {
        Task task = new Task(1, 15, 10, new ArrayList<>());
        List<Task> tasks = Collections.singletonList(task);
        int penalty = TaskScheduler.minPenalty(tasks);
        // Completion time: 15, deadline: 10, penalty: 5
        assertEquals(5, penalty, "Single task finishing late should incur penalty equal to lateness.");
    }

    // Test 4: Multiple Tasks with dependencies and on time
    @Test
    public void testMultipleTasksNoPenalty() {
        Task task1 = new Task(1, 10, 20, new ArrayList<>());
        Task task2 = new Task(2, 15, 30, Arrays.asList(1));
        Task task3 = new Task(3, 5, 40, Arrays.asList(2));
        List<Task> tasks = Arrays.asList(task1, task2, task3);
        int penalty = TaskScheduler.minPenalty(tasks);
        // Expected schedule: 1 (finishes at 10), 2 (finishes at 25), 3 (finishes at 30)
        // Penalties: 0, 0, 0 -> Total = 0
        assertEquals(0, penalty, "Valid dependency chain with tasks finishing on time should incur 0 penalty.");
    }

    // Test 5: Multiple Tasks with dependencies and penalties
    @Test
    public void testMultipleTasksWithPenalty() {
        // Task 1: finishes at 20, deadline = 15 => penalty 5
        // Task 2: depends on task 1, duration 10, scheduled after task 1, finishes at 30, deadline = 25 => penalty 5
        // Task 3: depends on task 1, duration 20, scheduled after tasks 1 and 2 in optimal ordering, finishes at 50, deadline = 60 => penalty 0
        Task task1 = new Task(1, 20, 15, new ArrayList<>());
        Task task2 = new Task(2, 10, 25, Arrays.asList(1));
        Task task3 = new Task(3, 20, 60, Arrays.asList(1));
        List<Task> tasks = Arrays.asList(task1, task2, task3);
        int penalty = TaskScheduler.minPenalty(tasks);
        // Optimal schedule minimizes penalty: expected total penalty = 5 + 5 = 10
        assertEquals(10, penalty, "Multiple tasks with penalties should compute the optimal total penalty.");
    }

    // Test 6: Task with zero duration
    @Test
    public void testZeroDurationTask() {
        Task task1 = new Task(1, 0, 0, new ArrayList<>());
        Task task2 = new Task(2, 10, 10, Arrays.asList(1));
        List<Task> tasks = Arrays.asList(task1, task2);
        int penalty = TaskScheduler.minPenalty(tasks);
        // Task1 finishes at 0 (deadline 0) penalty = 0; Task2 finishes at 10 (deadline 10) penalty = 0
        assertEquals(0, penalty, "Tasks with zero duration should be handled correctly, resulting in 0 penalty if on time.");
    }

    // Test 7: Complex dependency structure with minimal penalty
    @Test
    public void testComplexDependencies() {
        Task task1 = new Task(1, 5, 5, new ArrayList<>());
        Task task2 = new Task(2, 10, 15, Arrays.asList(1));
        Task task3 = new Task(3, 8, 20, Arrays.asList(1));
        Task task4 = new Task(4, 7, 30, Arrays.asList(2, 3));
        Task task5 = new Task(5, 12, 40, Arrays.asList(3));
        List<Task> tasks = Arrays.asList(task1, task2, task3, task4, task5);
        int penalty = TaskScheduler.minPenalty(tasks);
        // One optimal schedule:
        // Order: 1 (finishes at 5, penalty 0), 2 (finishes at 15, penalty 0), 3 (finishes at 23, penalty 3),
        // 4 (finishes at 30, penalty 0), 5 (finishes at 42, penalty 2) => Total penalty = 5
        assertEquals(5, penalty, "Complex dependency scheduling should compute the minimal total penalty correctly.");
    }

    // Test 8: Circular dependency should yield no valid schedule
    @Test
    public void testCircularDependency() {
        // Introducing a circular dependency: Task 1 depends on Task 2 and Task 2 depends on Task 1.
        Task task1 = new Task(1, 10, 20, Arrays.asList(2));
        Task task2 = new Task(2, 10, 20, Arrays.asList(1));
        List<Task> tasks = Arrays.asList(task1, task2);
        int penalty = TaskScheduler.minPenalty(tasks);
        // For circular dependency, no valid schedule exists so penalty should be Integer.MAX_VALUE.
        assertEquals(Integer.MAX_VALUE, penalty, "Circular dependencies should result in Integer.MAX_VALUE penalty.");
    }
}

File: cpp/exercises/practice/unit_test_5cc94226_99e9_4c4a_9444_1936e3640000/task_scheduler/task_scheduler_test.cpp
#include "catch.hpp"
#include "task_scheduler.h"
#include <vector>

using std::vector;

TEST_CASE("Single task, no dependencies and no penalty", "[task_scheduler]") {
    vector<int> duration = {10};
    vector<int> deadline = {20};
    vector<vector<int>> dependencies = {{}};
    int result = task_scheduler::min_total_penalty(duration, deadline, dependencies);
    // The task finishes at time 10 which is before the deadline of 20, so penalty is 0.
    REQUIRE(result == 0);
}

TEST_CASE("Chain dependencies with penalty", "[task_scheduler]") {
    // Tasks 0 -> 1 -> 2
    // Task0: duration 50, deadline 30 -> penalty = 20 as it finishes at 50.
    // Task1: duration 10, deadline 80 -> finishes at 60, penalty = 0.
    // Task2: duration 20, deadline 100 -> finishes at 80, penalty = 0.
    vector<int> duration = {50, 10, 20};
    vector<int> deadline = {30, 80, 100};
    vector<vector<int>> dependencies = {
        {},    // Task 0 has no dependencies.
        {0},   // Task 1 depends on Task 0.
        {1}    // Task 2 depends on Task 1.
    };
    int result = task_scheduler::min_total_penalty(duration, deadline, dependencies);
    // Expected total penalty = 20 (only task0 incurs penalty)
    REQUIRE(result == 20);
}

TEST_CASE("Multiple independent tasks with reordering", "[task_scheduler]") {
    // Three independent tasks where order can be optimized.
    // Task0: duration=30, deadline=25 -> penalty if delayed = (finish time - 25)
    // Task1: duration=10, deadline=50 -> penalty if finish > 50
    // Task2: duration=20, deadline=30 -> penalty if finish > 30
    // Optimal ordering gives total penalty = 35.
    vector<int> duration = {30, 10, 20};
    vector<int> deadline = {25, 50, 30};
    vector<vector<int>> dependencies = {
        {}, // Task0 has no dependencies.
        {}, // Task1 has no dependencies.
        {}  // Task2 has no dependencies.
    };
    int result = task_scheduler::min_total_penalty(duration, deadline, dependencies);
    REQUIRE(result == 35);
}

TEST_CASE("Parallel dependencies with merging", "[task_scheduler]") {
    // Graph structure:
    // Task0: no dependency, duration=10, deadline=10 -> finishes at 10, penalty=0.
    // Task1: depends on Task0, duration=20, deadline=25 -> if scheduled next, finishes at 30, penalty=5.
    // Task2: depends on Task0, duration=5, deadline=100 -> if scheduled after task0, finishes at 15, penalty=0.
    // Task3: depends on Task1 and Task2, duration=15, deadline=40 -> finishes after both are complete.
    // Optimal scheduling: Task0, Task1, Task2, Task3 results: Task0(10), Task1(30, penalty=5), Task2(35, no penalty), Task3(50, penalty=10) -> total = 15.
    // Alternatively, if Task2 is scheduled before Task1, penalty increases.
    vector<int> duration = {10, 20, 5, 15};
    vector<int> deadline = {10, 25, 100, 40};
    vector<vector<int>> dependencies = {
        {},      // Task0 no dependencies.
        {0},     // Task1 depends on Task0.
        {0},     // Task2 depends on Task0.
        {1, 2}   // Task3 depends on Task1 and Task2.
    };
    int result = task_scheduler::min_total_penalty(duration, deadline, dependencies);
    REQUIRE(result == 15);
}

TEST_CASE("Complex scenario with mixed dependencies", "[task_scheduler]") {
    // This test contains a more complex dependency graph:
    // Task0: duration=15, deadline=20
    // Task1: duration=25, deadline=50, depends on Task0.
    // Task2: duration=10, deadline=25, depends on Task0.
    // Task3: duration=20, deadline=60, depends on Task1 and Task2.
    // Task4: duration=30, deadline=100, depends on Task2.
    // Expected schedule example:
    // - Task0: [0,15] => penalty = max(15-20, 0) = 0.
    // - Task2: [15,25] => penalty = max(25-25, 0) = 0.
    // - Task1: [25,50] => penalty = max(50-50, 0) = 0.
    // - Task3: [50,70] => penalty = max(70-60, 10) = 10.
    // - Task4: [70,100] => penalty = max(100-100, 0) = 0.
    // Total penalty expected = 10.
    vector<int> duration = {15, 25, 10, 20, 30};
    vector<int> deadline = {20, 50, 25, 60, 100};
    vector<vector<int>> dependencies = {
        {},      // Task0 no dependencies.
        {0},     // Task1 depends on Task0.
        {0},     // Task2 depends on Task0.
        {1, 2},  // Task3 depends on Task1 and Task2.
        {2}      // Task4 depends on Task2.
    };
    int result = task_scheduler::min_total_penalty(duration, deadline, dependencies);
    REQUIRE(result == 10);
}
  
// Main runner for catch if needed to compile tests as standalone
#define CATCH_CONFIG_MAIN
#include "catch.hpp"